token_contracts/baln/balance.py

```diff
@@ -522,6 +532,138 @@ def burnFrom(self, _account: Address, _amount: int) -> None:
+            self._total_stake_snapshot[current_id]['amounts'][length] = current_value
+            self._total_stake_snapshot[current_id]['length'][0] += 1
+            return
+        else:
+            last_snapshot_id = self._total_stake_snapshot[current_id]['ids'][length - 1]
+
+        if last_snapshot_id < current_id:
+            self._total_stake_snapshot[current_id]['ids'][length] = current_id
+            self._total_stake_snapshot[current_id]['amounts'][length] = current_value
+            self._total_stake_snapshot[current_id]['length'][0] += 1
+        else:
+            self._total_stake_snapshot[current_id]['values'][length - 1] = current_value
+
+    @external(readonly=True)
+    def stakedBalanceOfAt(self, _account: Address, _day: int) -> int:
+        if _day < 0:
+            revert(f'{TAG}: snapshot id is equal to or greater then Zero')
+        low = 0
+        high = self._stake_snapshot[_account]['length'][0]
+
+        while low < high:
+            mid = (low + high) // 2
+            if self._stake_snapshot[_account]['ids'][mid] > _day:
+                high = mid
+            else:
+                low = mid + 1
+        if self._stake_snapshot[_account]['ids'][0] == _day:
+            return self._stake_snapshot[_account]['amounts'][0]
+        elif low == 0:
+            return 0
+        else:
+            return self._stake_snapshot[_account]['amounts'][low - 1]
+
+    @external(readonly=True)
+    def totalStakedBalanceOfAt(self, _day: int) -> int:
+        if _day < 0:
+            revert(f'{TAG}: snapshot id is equal to or greater then Zero')
+        low = 0
+        high = self._total_stake_snapshot[_day]['length'][0]
+
+        while low < high:
+            mid = (low + high) // 2
+            if self._total_stake_snapshot[_day]['ids'][mid] > _day:
+                high = mid
+            else:
+                low = mid + 1
+        if self._total_stake_snapshot[_day]['ids'][0] == _day:
+            return self._total_stake_snapshot[_day]['amounts'][0]
+        elif low == 0:
+            return 0
+        else:
+            return self._total_stake_snapshot[_day]['amounts'][low - 1]
+
+    @external
+    @only_owner
+    def loadBalnStakeSnapshot(self, _data: bytes) -> None:
+        d = json_loads(_data.decode("utf-8"))
+        _days = list(d.keys())
+        for current_id in _days:
+            if int(current_id) < self.getDay():
+                for _id in d.get(current_id):
+                    _account = Address.from_string(_id.get('address'))
+                    length = self._stake_snapshot[_account]['length'][0]
+                    self._stake_snapshot[_account]['ids'][length] = int(current_id)
+                    self._stake_snapshot[_account]['amounts'][length] = int(_id.get('amount'))
+                    self._stake_snapshot[_account]['length'][0] += 1
+            else:
+                pass
+
+    @external
+    @only_owner
+    def loadTotalStakeSnapshot(self, _data: bytes) -> None:
```

Comment:
Please use List to take input rather than bytes and doing json.loads. It would increase the transaction fee.

Reply:
updated to TypedDict type.